contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -274,6 +279,329 @@ transform_funcexpr(Node *node)
+										InvalidOid, server_collation_oid, oprfuncid(optup));
+		ReleaseSysCache(optup);
+		/* construct leftop < pattern */
+		optup = compatible_oper(NULL, list_make1(makeString("<")), ltypeId, ltypeId,
+								true, -1);
+		if (optup == (Operator) NULL)
+			return node;
+
+		less_equal = make_op_with_func(oprid(optup), BOOLOID, false,
+										(Expr *) leftop, (Expr *) concat_expr,
+										InvalidOid, server_collation_oid, oprfuncid(optup));
+		constant_suffix = make_and_qual((Node *) greater_equal, (Node *) less_equal);
+		if (like_entry.is_not_match)
+		{
+			constant_suffix = (Node *) make_notclause((Expr *) constant_suffix);
+			ret = make_or_qual(node, constant_suffix);
+		}
+		else
+		{
+			constant_suffix = make_and_qual((Node *) greater_equal, (Node *) less_equal);
+			ret = make_and_qual(node, constant_suffix);
+		}
+		ReleaseSysCache(optup);
+	}
+	return ret;
+}
+
+/*
+ * Function responsible for obtaining unaccented version of input
+ * string with the help of ICU provided APIs. 
+ * We use a transformation rule to transliterate the string
+ */
+
+PG_FUNCTION_INFO_V1(remove_accents);
+Datum
+remove_accents(PG_FUNCTION_ARGS)
+{
+	text *arg1 = PG_GETARG_TEXT_PP(0);
+	char *input = text_to_cstring(arg1);
+	UChar *rules; // Rule to remove diacritics 
+	UChar *utf16_intput;
+	int32_t len_uchar,
+			len_uinput,
+			limit,
+			capacity,
+			len_result;
+	UErrorCode status = U_ZERO_ERROR;
+	char *result;
+	UTransliterator *transliterator;
+
+	/* Rule applied to transliterate Latin and general category Nd character 
+	 * then convert the Latin (source) char to ASCII (destination) representation
+	 */
+	char* rule = "[[:Latin:][:Nd:]]; Latin-ASCII";
+
+#ifdef USE_ICU
+	len_uchar = icu_to_uchar(&rules, rule, strlen(rule));
+	len_uinput = icu_to_uchar(&utf16_intput, input, strlen(input));
+	
+	// Open a transliterator for conversion
+	transliterator = utrans_openU(rules, len_uchar, UTRANS_FORWARD, NULL, 0, NULL, &status);
+	if (U_FAILURE(status))
+	{
+		elog(ERROR, "Error opening transliterator: %s", u_errorName(status));
+		return 1;
+	}
+
+	limit = len_uinput;
+	capacity = VARCHAR_MAX_SCALE;
+
+	utrans_transUChars(transliterator,
+						utf16_intput,
+						&len_uinput,
+						capacity,
+						0,
+						&limit,
+						&status);
+
+	if (U_FAILURE(status))
+	{
+		elog(ERROR, "Error normalising the input string: %s", u_errorName(status));
+	}
+
+	len_result = icu_from_uchar(&result, utf16_intput, len_uinput);
+
+	PG_RETURN_VARCHAR_P(cstring_to_text_with_len(result, len_result+1));
+#endif
+	PG_RETURN_NULL();
+}
+
+static Node *
+ConvertNodeToFuncExpr(Node *node, void *context)
+{
+	FuncExpr *newFuncExpr = makeNode(FuncExpr);
+	Oid funcargtypes[1] = {CHAROID};
+	newFuncExpr->funcid = LookupFuncName(list_make1(makeString("remove_accents")), -1, funcargtypes, true);;
+
+	if (node == NULL)
+		return node;
+
+	switch (nodeTag(node))
+	{
+		case T_Const:
+			{
+				Const *con = (Const *) node;
+				newFuncExpr->funcresulttype = con->consttype;
```

Comment:
As remove_accents (remove_accents_internal) takes text as input, it can accept any string datatype. And we are already maintaining the return type by matching the return type of the New FuncExpr node